pilot/pkg/bootstrap/server.go

```diff
@@ -439,6 +440,11 @@ func (s *Server) Start(stop <-chan struct{}) error {
 		log.Warn("Server is starting with unsafe features enabled")
 	}
 
+	// run client before other components started in case other components need to run after specific cache sync
+	// It is to run informers created in component init function
+	if s.kubeClient != nil {
```

Comment:
I don't get why we need this. There is no change to when informers are created in this PR, no?

Reply:
It is needed, because we need to start the informer initialted in component init function. So we can run discovery namespace filter with namespace cache synced. Other components may also need specific informer synced before run.

Reply:
components should do

```
run() {
  wait for sync
  .. do whatever.. 
}
```

This pattern works fine today since they will wait.

Reply:
not all works, because the current client run at last

Reply:
namespace filter is the exception

Reply:
Why is initMulticluster not doing `Run()` in a goroutine? That seems like the real bug here, and we are just working around it.

I guess it was doing most of its work in a goroutine now, and we have just made in this PR `SyncNamespaces` required to complete before others. But this doesn't seem like the right fix - I think this goes back to https://github.com/istio/istio/issues/44675. We should start the filter as "nothing is allowed, HasSynced=false", all dependant informers are also "empty, HasSynced=false". Then when namespaces are synced we will eventually have our proper state and all dependent informers will be notified. It does not need to be synchronous.

IMO this is introducing behavior that is counter to our architecture and possibly dangerous, in order to work around more bugs in a niche part of the code that is fundamentally broken in a variety of ways. I think a proper fix on https://github.com/istio/istio/issues/44675 would be advantageous and less risky.

1.18 is also already shipped, so we have 3 months until next release to get it right

Reply:
>Why is initMulticluster not doing Run() in a goroutine?

This is a good idea. We can move the init into startfunc as well

Reply:
> > Why is initMulticluster not doing Run() in a goroutine?
> 
> This is a good idea. We can move the init into startfunc as well

As far as I can tell the main work of initMulticluster ends up calling in a goroutine inside the function. But the new code in this PR is done outside the goroutine so it blocks

Reply:
hmm,wait  you are mainly to resolve the bug, however the initial intent of the pr is to init namespace filter before running

Reply:
@howardjohn Lets resolve https://github.com/istio/istio/issues/44675 separately